tests/upgrade/test_upgrade_charm.py

```diff
@@ -0,0 +1,78 @@
+import logging
```

Comment:
I think we can leave the tests directory flat, i.e. move this up to be under `tests/`. Since pytest offers pretty powerful filtering via `-k`, I'm not worried about organizing the tests themselves into subdirectories.